src/order/conditionally_complete_lattice.lean

```diff
@@ -769,3 +769,182 @@ instance (α : Type*) [conditionally_complete_linear_order α] :
   ..order_dual.decidable_linear_order α }
 
 end order_dual
+
+section with_top_bot
+
+open lattice
+
+open_locale classical
+/-- Extends `Sup` from a conditionally complete lattice `α` to `with_top α`.
+The new `Sup` returns a non-junk value for more subsets (e.g. for unbounded-above subsets of `α`). -/
+noncomputable instance with_top.conditionally_complete_lattice.has_Sup
+  {α : Type*} [conditionally_complete_lattice α] : has_Sup (with_top α) :=
```

Comment:
You just need `has_Sup α` to define `has_Sup (with_top α)`. Same below for `Inf`.

Reply:
I completely agree, but when I tried this I discovered that on line 685 there was already an instance which clashed with this. What do I do?

Reply:
```
@[instance]
protected noncomputable def with_top.lattice.has_Sup : Π {α : Type u} [_inst_1 : conditionally_complete_linear_order_bot α], has_Sup (with_top α) :=
λ {α : Type u} [_inst_1 : conditionally_complete_linear_order_bot α],
  {Sup := λ (s : set (with_top α)), classical.some _}
```

Reply:
and the definition is completely different to mine (although it might well be provably equal)

Reply:
Have I made a diamond? I would appreciate advice on this one.

Reply:
Yes, you have created a diamond, and this is bad. What you should do is register your instance for `has_Sup (with_top α)` under the assumption `has_Sup α`, before line 685, and then prove the results around line 685 and your results using this same definition.

Reply:
OK this makes a lot of sense. I'll give it a go!

Reply:
So I was a bit scared that this was a high risk strategy, because this file is imported by all sorts of other things e.g. data.real.basic and some algebra stuff -- but I redefined has_Sup (with_top alpha) and the Inf version, and have managed to get everything compiling!

Reply:
I have the impression that you are doing twice (or three times) the same thing. The proof of `is_lub_Sup` (showing that `Sup s` is a least upper bound of `s` in `with_top alpha` where `alpha` is a `conditionally_complete_linear_order_bot`) should work in any conditionally complete lattice -- then you can use it also in the proof that `with_top alpha` is conditionally complete when `alpha` is, and also in the proof that `with_top (with_bot alpha)` is complete. Does that make sense?

Also, `has_lub` (in the section on `conditionally_complete_linear_order_bot`) is now useless, so you can remove it.

Reply:
The proof that `Sup s` is a least upper bound of `s` does not work for `with_top alpha` with `alpha` a conditionally complete lattice, because it's not true for the empty set (Sup returns junk in this case). This is precisely the reason the axiom is added in the definition of `conditionally_complete_linear_order_bot` I guess. And now an annoying issue is that if `alpha` is conditionally complete then `with_bot alpha` might not be a `conditionally_complete_linear_order_bot` because `alpha` might not be linearly ordered :-( This goes back to the issue not having a type `conditionally_complete_lattice_with_bot`, which would be an axiomatised version of `with_bot <conditionally complete lattice>`. 

My fix is to define a new function `is_lub_Sup' :
  ∀ {β : Type u_1} [_inst_2 : conditionally_complete_lattice β] (s : set (with_top β)),
    s ≠ ∅ → is_lub s (Sup s)` 

which is where the work is, and now everything else follows from this (I did indeed add one proof and then delete two). I have pushed what I have so far. I have not thought about whether a similar trick will work for Inf.

I've also removed the `has_lub` and `has_glb` code. The functions were not used anywhere else in mathlib.

Reply:
I guess the analogous fix for Inf is to show that if alpha is a conditionally complete lattice then for any (possibly empty) bounded-below subset of with_top alpha, Inf returns the correct glb. I can then use this one proof to replace the code at 710-733 (because all sets are bounded below) and 843-869 (because bounded below is an assumption here), if I've understood correctly. I've not yet done this so this PR remains unfinished. I'll try to do it later today.

Reply:
OK I did it! We now have `is_glb_Inf'` too, and this one proof replaces not two but *three* proofs in fact, because I can also use this and `is_lub_Sup'` in the final proof. OK I think I have finally dealt with this issue Sebastien. The only other thing I could envisage is making a `conditionally_complete_lattice_with_bot` class instead of proving things about `with_bot <conditionally_complete_lattice>` and `conditionally_complete_linear_order_bot`; I am honestly not sure that this is worth it but I'll do it if you think it's a good idea.